### PR TITLE
Migrate to Ethers

### LINKED ISSUE #143: Move to ethers.js in Plasma Cash bindings
Currently, the Plasma Cash bindings are all using web3.js for interacting with mainnet. We should move to Ethers.js which has strictly better APIs for monitoring events, as well as creating transactions offline. This also allows us to have both offline tx signing and metamask in a very intuitive way.

This will be implemented by wrapping the `createUser` function from the `metamask` branch as follows:

```
User.newMetamaskUser(<same args as createUser without web3>) {
  const provider = new ethers.providers.Web3Provider(web3.currentProvider);
  const wallet = provider.getSigner();
  return User.createUser(wallet, ...args)
}
```

The above will create a net `User` instance which is instantiated to work with metamask.

The offline user will be done with:
```
User.newOfflineUser(privateKey, endpoint, <same args as createUser without web3>) {
  const privateKey = <yourpk>
  const provider = new ethers.providers.JsonRpcProvider(endpoint)
  const wallet = new ethers.Wallet(privateKey, provider);
  return User.createUser(wallet, ...args)
}
```

This will require changing the `Entity` class to accept a ethers.js Signer object instead of web3, as well as any necessary API changes to the contracts for calling methods and listening to events.

---

Below I am adding some example usages of ethers.js, along with docs:
https://docs.ethers.io/ethers.js/html/cookbook.html?highlight=cookbook

# Connecting with a Privatekey:
```
var ethers = require('ethers');

var privateKey = <yourpk>
var provider = ethers.providers.getDefaultProvider('rinkeby');
var wallet = new ethers.Wallet(privateKey, provider);

var address  = <contractaddress>
var abi = <yourabi>

var contract = new ethers.Contract(address,abi,wallet);

await contract.someFunction(<args>)
```

# Connecting to metamask:
```
var ethers = require('ethers');
const provider = new ethers.providers.Web3Provider(web3.currentProvider);
const wallet = provider.getSigner();

var address  = <contractaddress>
var abi = <yourabi>

var contract = new ethers.Contract(address,abi,wallet);

```

## Sending ETH
```
 let tx = await wallet.sendTransaction({
        gasLimit: gasLimit,
        gasPrice: gasPrice,
        to: newAddress,
        value: value
    });
    let receipt = await tx.wait()
```

##  Sending ETH along a function 
```
let tx = contract.someFunction(addr, { value: utils.parseEther('1.0') })
let receipt = await tx.wait()
```

### Listening for all occurences of an event
```
contract.on("ValueChanged", (author, oldValue, newValue, event) => {
    // Called whenever ValueChanged is emitted. The callback arguments are the event params.
    console.log(author);
    // "0x14791697260E4c9A71f18484C9f997B308e59325"
    console.log(oldValue);
    // "Hello World"
    console.log(newValue);
    // "Ilike turtles."
    // See Event Emitter below for all properties on Event
    console.log(event.blockNumber);
    // 4115004
});
```

### Filtering

```
// The null field indicates any value matches
// This will filter all Transfer events where the second argument is myAddress.
let filter = contract.Transfer(null, myAddress);

// Listen for our filtered results
contract.on(filter, (from, to, value) => {
    console.log('I received ' + value.toString() + ' tokens from ' + from);
});
```

### PR DESCRIPTION
**Pull request by gakonst**
Closes #143 

- Removes SignedContract in favor of utilizing Ethers for both offline and metamask compatible signing. 
- `createMetamaskUser` and `createOfflineUser` functions added to `User` class.
- Listening to events from Metamask is a pain. We add a second instance of the Plasma Cash contract (`plasmaEvents`) which is hooked up to a websocket endpoint to listen for events. Making this worth with ethers involved significant amount of engineering and I could not make it work in the same way as before, so I went with a Web3 contract instance here.
- Minor changes to the PlasmaDB.ts
- Merges with #149 and updates e2e tests as they are in https://github.com/loomnetwork/plasma-cash/pull/160
This PR gives us a lot of flexibility regarding how users will be able to initialize a Plasma Cash account instance. However, there is the added overhead of maintaining both Web3+Ethers. A future PR will follow which will make the Plasma Cash event listener contract with an Ethers based one, eliminating the dependency of Web3.js

### CODE REVIEW

File: src/plasma-cash/ethereum-client.ts [hunk 1]
```diff
@@ -206,19 +221,25 @@ export class EthereumPlasmaClient {
...
-        slot.toString(),
-        prevTxBytes,
-        exitTxBytes,
-        prevTx ? prevTx.proof : '0x',
-        exitTx.proof,
-        exitTx.sig,
-        [prevBlockNum ? prevBlockNum.toString() : 0, exitBlockNum.toString()]
-      ],
-      bond
+      '0x' + slot.toString(16),
```
Comment (enlight) [inline src/plasma-cash/ethereum-client.ts]: Not loving this `'0x' + slot.toString(16)` everywhere, does `slot.toString()` not work anymore?

File: src/plasma-cash/user.ts [hunk 2]
```diff
@@ -13,14 +13,16 @@ import {
...
 } from '..'
 import { IPlasmaCoin } from './ethereum-client'
-import { sleep } from '../helpers'
-import { IWeb3EventSub } from './entity'
+import { sleep, hexBN } from '../helpers'
+import { ethers, providers } from 'ethers'
 
 const ERC721 = require('./contracts/ERC721.json')
+
+const ERC721_ABI = ['function safeTransferFrom(address _from, address _to, uint256 _tokenId)']
```
Comment (enlight) [inline src/plasma-cash/user.ts]: Is this supposed to replace `ERC721.json`? Can we get rid of the json file then?

File: src/tests/e2e/address-mapper-tests.ts [hunk 3]
```diff
@@ -10,12 +10,14 @@ import {
...
 import { Web3Signer } from '../../solidity-helpers'
+import { ethers } from 'ethers';
 
 const Web3 = require('web3')
 
 // TODO: Need a factory to create connection properly likes plasma-cash test
 function getWeb3Connection() {
-  return new Web3('http://127.0.0.1:8545')
+  const provider = new ethers.providers.JsonRpcProvider('http://localhost:8545')
+  return provider.getSigner()
```
Comment (enlight) [inline src/tests/e2e/address-mapper-tests.ts]: Can you renamed this method to something more appropriate now that it doesn't actually return a Web3 instance?

File: src/solidity-helpers.ts [hunk 4]
```diff
@@ -20,16 +20,14 @@ export interface IEthereumSigner {
  * This signer should be used for interactive signing in the browser with MetaMask.
  */
 export class Web3Signer implements IEthereumSigner {
```
Comment (enlight) [inline src/solidity-helpers.ts]: Can you create a new `EthersSigner` and leave the existing `Web3Signer` and `OfflineWeb3Signer` in place for now and just add a comment to them saying they're deprecated. We can remove them when we've gotten rid of the Web3 dependency entirely, until then they're still used in some other projects.

File: src/plasma-cash/ethereum-client.ts (hunk 1 above)
Comment (gakonst) [reply src/plasma-cash/ethereum-client.ts]: Yeah unfortunately it doesn't, this is an under the hood difference between web3/ethers

File: src/plasma-cash/user.ts (hunk 2 above)
Comment (gakonst) [reply src/plasma-cash/user.ts]: Yes

File: src/tests/e2e/address-mapper-tests.ts (hunk 3 above)
Comment (gakonst) [reply src/tests/e2e/address-mapper-tests.ts]: Yes

File: src/solidity-helpers.ts (hunk 4 above)
Comment (gakonst) [reply src/solidity-helpers.ts]: Yes

**Comment by enlight** [comment]
Alrighty then, LGTM

**Pull request merged**